### DISCUSSION TITLE
Potential proposal for updating the delay falloff (or similar) workflow

### DISCUSSION THREAD

**Discussion by dapa5900** [JacquesLucke/animation_nodes#1760] [Ideas]
Hi @OmarEmaraDev 

as you know I'm currently working on this book generator setup and I have realized that I can not really figure out a good way how to procedurally flip the pages while still having full control over to which pages exactly I can flip but also control the offset / the delay of the pages in between the flips (meaning to have multiple pages flip at once with slight time offsets instead of doing full page per page flips). As you might know from my setup I've tried a very hacky way by essentially keyframing the time and duration parameters of the time delay node manually which sort of works if you're lucky to hit a sweet spot but can also result in very wonky animations as you essentially changing the speed of a flip throughout the animation. And this completely falls apart when you try to flip to more than one page one after another as by changing the duration again the book flips back to previous pages as well. I also talked to Bradley about this and he came up with an interesting fallback approach by utilizing a controller falloff but I'm right now still wondering if there might be already a good fully procedural way to do this?

To break this down to a simpler setup I've created this example with manual key-framing:

https://user-images.githubusercontent.com/75205162/127128813-64f00daa-953c-4e91-80ab-b6e65134a3a0.mp4

So the whole thing is basically about affecting instances similar like you can do this via the delay falloff but also be able to control exactly which instances are gonna be affect, where to have breaks in the animation, where to continue in time and bonus, also have control about different timings/interpolation of the offsets (as you can see from the keyframing: first part has 1 frame offset between each object, second 2 frames and the last part as a curve offset). You can also finde the file and video here: https://p7s1-my.sharepoint.com/:f:/g/personal/par0001d_belgium_fhm_de/EuitBldqj9JKj9odtN5iQt4BlHyVGN5N-g1koPWxoDdH-g?e=dHO6ll

My question is, if there's already a good way to this in AN? And if not I have some ideas I'd like to propose to you to maybe add a "Manual" mode to the delay effector to achieve these kind of results.

Thanks!

**Comment by OmarEmaraDev** [comment]
The Delay Falloff is one of the nodes that I want to see improved, so any suggestions there is most welcome regardless.

I don't have much time to think of a good way to do what you want, but I will give you some examples for inspiration. All of the examples utilize the input Offsets of the Delay Falloff node to manually control the offsets of the motion. Giving no Offsets list is equivalent of giving an arithmetic sequence starting at zero with a difference of one, that is, `[0, 1, 2, 3, ...]`. Lets call this the base offset.

### Constant Groups Constant Delay

If you want to animate every consecutive `n` objects together with a delay of `m` between groups, you take the floor division by `n` then multiply by `m` and add to the base offset. The following example have `n = 3` and `m = 5`.

https://user-images.githubusercontent.com/25300994/127145905-48e8af2a-eb5b-4ad8-a963-51f872c86ebf.mp4

### Constant Group Variable Delay

If you want to manually specify the delay between each group, specify the offset of each group in a float list and repeat the elements of the list by the number of elements in the group. The following examples makes the third and fourth group close to each other and the last group distant from the rest.

https://user-images.githubusercontent.com/25300994/127146595-08179d53-4a15-4613-9b65-bc9aa364cbe0.mp4

### Variable Group Variable Delay

If you also want to specify the number of elements in each group, you can have another float list storing that and use it as the amount of repetition. The following example specify a group of 3, followed by 5, followed by 1, followed by 6.

https://user-images.githubusercontent.com/25300994/127147069-47a5c0bd-6e12-4f5d-ab10-3d6bb45f73d5.mp4

### Variable Group Variable Delay Variable Interpolation

We can't control the interpolation directly in the Delay Falloff, so you can set it to Linear and interpolate the falloff after, either by switching the interpolation or by using a mask falloff. Let me know if you need more details about this.

https://user-images.githubusercontent.com/25300994/127148126-dbcdea58-d57b-427e-9d8d-3ad4291184a0.mp4

**Reply by dapa5900** [reply]
@OmarEmaraDev 

thank you for these example and I really appreciate your effort for putting this together. 

I rebuilt this on my side and have two questions for now. 

1. Is there an effective way to do a calculation to specifically set the exact starting frames for each animation block because currently the actual timing is very much dependent on the delay, duration and instances amount numbers. I tried to do manual calculations for this (see below) but this results in a lot of manual work, maybe there's an easier way to do this?

![frameStarts](https://user-images.githubusercontent.com/75205162/127287396-987d68e5-b58b-4d54-884d-17e0e69d422a.png)

2. Is there a way to reverse these animations but then without the gaps in between? Think about the book generator, where you would flip from cover/page 0 to eg page 32, pause, page 54, pause, page 72 pause and then flip back from page 72 to the cover/page 0 without any pauses in between?

I'm currently also pretty busy, but will come up with a proposal for the delay effector soonish.

Thanks!

**Reply by OmarEmaraDev** [reply]
@dapa5900

1. There are more automated ways, but they are not easy either. So I am afraid you will have to do this computation one way or the other.
2. Not sure if this is what you want, but reversing the offsets will give you want you want. You can switch the falloff at the right time to combine both forward and backward ones.
  
https://user-images.githubusercontent.com/25300994/127363891-23a392ee-a832-420c-b29f-1806687a3e20.mp4

**Reply by dapa5900** [reply]
@OmarEmaraDev 

I have one more question regarding this (at least I hope so haha ;)). Is there a way to extract the number of the currently affected instance from this setup somehow, basically the progress of the delay effector? Bradley helped me on his discord already with this and suggested this method:

![instanceNumber](https://user-images.githubusercontent.com/75205162/127998686-52c389b8-6178-492a-a002-feb6211cfb43.png)

This is already working very well however it only gives me integers, so while the flipping animation is running, the numbers jump from 1 to 2 to 3, etc. Since I'm using these values to driver other things in the setup as well it would be nice to somehow extract the exact progress as floats to have more granular values available because even though very subtle with this approach the flipping animations also have a slightly jagged feeling. (in my old setup I used a slider with a float value to drive this manually so i also got in between numbers like 1.32, 2.67, etc.)

Thank you!

**Reply by OmarEmaraDev** [reply]
@dapa5900 Use the *Float List Math* node to sum the output of the *Evaluate Falloff* node, then divide the sum by the the number of pages. 0 will mean no pages are turned, 1 means all pages are turned.

**Reply by dapa5900** [reply]
@OmarEmaraDev 

works like a charm :) thanks a lot!

**Comment by dapa5900** [comment]
Hi @OmarEmaraDev 

I finally managed to setup the book generator with all the features I would like to have. I used material instancing for now instead of the shader selector addon and render with cycles. It works and looks good in the viewport but when rendering the scene I get two issues:

1. The animations looks quite jaggy; overall less smooth than in the viewport and sometimes frames get duplicated (eg Frame 79 and 80). I tried serveral things to solve this including disabling different features and baking the animation down to keyframes via "Set Keyframes" node. The last option seems to help however I only managed to bake down LocRotScale of the objects but not the bend angles, etc. from the modifiers (basically anything that I set via Object Attribute Output and tried to bake via a custom keyframe property...It says "Could not create Keyframes")
2. I get a lot of crashes, sometimes earlier, sometimes later. I tried as well different things, including different render settings and also rendering via CPU instead of GPU but so far I never managed to render the whole timeline start to finish.

Here you can find the latest project version, a preview of the render and a couple of crash reports from my different trials: 

https://p7s1-my.sharepoint.com/:f:/g/personal/par0001d_belgium_fhm_de/EoK4mjbvqxJJjh09hYCcmQIB2lWGAJjUqSF3uQQkT-keHw?e=dqI50w

Thank you!

**Reply by OmarEmaraDev** [reply]
@dapa5900 Locking the UI during rendering doesn't solve the crashes?

**Reply by dapa5900** [reply]
@OmarEmaraDev 

just tried it, still crashes. the report is called "MeinHerzenspferd_08.crash.txt"

**Reply by OmarEmaraDev** [reply]
@dapa5900 Blender is on a crash spree lately ... :disappointed:. In this particular case, Blender is crashing while drawing the image editor. So just close the image editor as a workaround for now. It shouldn't draw it when the UI is locked anyway ...

**Reply by dapa5900** [reply]
@OmarEmaraDev 

hm, still crashing here, even sooner now :( new crash report: "MeinHerzenspferd_09.crash.txt"

**Reply by Gerstmann-Bradley** [reply]
i think timing of crash doesn't matter since it's random. I tested file in eevee it's fine for me once locked interface. Not interested in testing in cycle.
Nevertheless, if it's bad in your case, either you render piece by piece, or perhaps the script method is your last option.
I personally no longer use nor keep the script. Perhaps Omar can provide it again?

**Reply by dapa5900** [reply]
@OmarEmaraDev

thanks. I now also closed the shader editor and animations nodes editor and now it worked. this also resolved the issue with the jagged animation in the render obliviously, now it looks smooth :). as a workaround this works, don't know if this interface update issue should be reported as bug to Blender directly...?

@Gerstmann-Bradley 

thanks for testing the file. do the textures render properly for you in eevee? I mean having different textures per page? for me this only works in cycles...

**Reply by OmarEmaraDev** [reply]
The script is available in #1200. It should be reported to Blender, but a report in the current form will probably be missed. We will have to make an effort to make a minimal file to replicate the issue. I will try to do that and let you know how it goes.

**Reply by Gerstmann-Bradley** [reply]
your texture isn't packed, so I can't see them other than pinkish replacement.
I haven't checked your node tree.
But upon guessing the issue with eevee is how it handles alpha. Other than that, ideally it should have no problem.

**Reply by Gerstmann-Bradley** [reply]
ok I see your problem in eevee. I am not sure how to comment on this, but if in current state if this works in cycle it's good. At the end I don't think eevee and cycle should matter in terms of crash. I personally don't wish to try though.
